src/java.base/share/classes/sun/security/ssl/ServerHello.java

```diff
@@ -406,6 +403,7 @@ private static KeyExchangeProperties chooseCipherSuite(
                 ClientHelloMessage clientHello) throws IOException {
             List<CipherSuite> preferred;
             List<CipherSuite> proposed;
+
             if (shc.sslConfig.preferLocalCipherSuites) {
                 preferred = shc.activeCipherSuites;
                 proposed = clientHello.cipherSuites;
```

Comment:
Instead of wrapping all information in one block, I may prefer to break down and place them in the close place where is happens.

I may dump the debug log here for cipher suite preference and server activated cipher suites.

Reply:
The intention to group all the information together allows the developer to find the configuration of the server socket in one place. Previously the log message will only print out the error message of "on cipher suite in common" when the handshake failed between the client and server.

Reply:
Intend to log all of the collected information in one place. It is easy for the developer to see the configuration of the server socket. Previously when the handshake is failing, no information is available for the server socket in the log file.

Reply:
So, do you want to make the log where the configuration happens?  Logging in one place cannot have the accuracy debug log where the problem happens.  One just get the information, but did not get the code line numbers and processes why the information is not good.  We can have all log in one place for the TLS implementation, but it was not choose because more debug information was expected to carry in the log.

Reply:
I think Weibing is trying to achieve a balance here - the current TLS logs are quite verbose. I'm not sure if we need verbose SSLServerSocket info for every server socket operation. The current approach is to print the SSLServerSocket details when a handshake fails due to a ciphersuite/keyexchange config issue

Reply:
Thanks for the comments.  I'm not sure if it is really helpful for developers to understand and debug the failure by reading the additionally dumped cipher suites and/or key exchange configuration.  Given the server cipher suites TLS_AES_128_GCM_SHA256, can one really know the failure reason exactly?

Reply:
The cipher suite enabled on the server side is not logged when "no common in cipher suites" error is thrown. Hope the developer could find the difference in the cipher suites between client and server.

Reply:
Even the cipher suites are the same between client and server, it may still fail with "no common in cipher suites" error.  The cause of the bug is not only about "no common in cipher suites" between client and server, but also about the cases that the server cannot negotiate any of the common cipher suites because of other facts.  If the cause is only about ""no common in cipher suites", I don't think we need a fix as the message has been tell the story.

It is not objected that cipher suite should not be logger.  The question raised here is about when and how to log the cipher suite, and how to make the log easier to read and easier to debug.

Reply:
@XueleiFan  - I think it's fair to say that the current "no cipher suites in common" exception message is misleading for some scenarios. If not misleading, it's ambiguous. You could be dealing with a simple scenario where the server socket simply has no enabled cipher suites to match those coming in from the ClientHello. In the more complex scenario, something else has caused the fatal issues like a key exchange issue in TLSv1.2.  With the new debug data, we have some extra data points like the ciphersuites actually enabled, the preference of client or server suites etc. Coupled with the ClientHello data, I think it helps piece together a better idea of the server side scenario.

I've no strong preference on how often we print this debug data on server end. To cut back on log verbosity, perhaps on fatal/failing handshake scenarios is all that's required. 

If you feel strongly about printing debug data for each TLSServerSocket handshake, (including the extra data points that you've suggested) then please shout now. I do feel it'll add considerably to log size on busy TLS Servers.

Reply:
@coffeys Before we go further, I think we'd better have a look if the information added could help for debugging.  I'm not sure of that now.

> With the new debug data, we have some extra data points like the ciphersuites actually enabled, the preference of client or server suites etc. Coupled with the ClientHello data, I think it helps piece together a better idea of the server side scenario.

If I understand correctly, the preference of client or server suites should not impact the success of handshaking.  Did you see examples otherwise in practice?   What's the purpose by adding this information for failed handshaking?

The client authentication type is also logged for failed server hello handshake message.  What's the purpose for the adding?  Did you see server hello failure that caused by client authentication type in practice?

With server debug log enabled, there have already been a lot debug information, especially for verberos debug mode.  Is it a really concern to you to dump server enabled cipher suites, just as the dump of ClientHello message?  Yes, it make debug logger worse for a busy server, but the impact may be 1/10K?

I have to run for now, and will try to reply more for your comments.  Thanks!

Reply:
> I've no strong preference on how often we print this debug data on server end. To cut back on log verbosity, perhaps on fatal/failing handshake scenarios is all that's required.

Does the switch on/off the "verbose", "data" or "record" debug option help to reduce the log verbosity?

In general, I'm not sure how on failing handshake scenarios could help to make the debug better.  Thinking about the failure of the ServerHello in this case, it could be related to ClientHello, named groups/key exchange algorithms, signature algorithms, key store, certificate signature, certificate key algorithms, certificate key length, etc.  If these information does not dump when the ClientHello is good, I'm not sure if the developer understand better about what could be the cause from the server cipher suite preference, client authentication types and server enabled cipher suites.  Well, we can dump everything in the printServerSocketConfig() method, but do we know everything in the context of printServerSocketConfig and is it easy to maintain the things in the method.  I'm not sure of that.

Reply:
I think many engineers use the `-Djavax.net.debug=all` option for TLS debugging at moment. `-Djavax.net.debug=ssl` is pretty much broke. There's an open bug for that IIRC.

I revisited the ServerHello produce() code for TLSv12/TLSv13 to see if this approach was making sense. In general, I think it's a good addition to the debug logs. They're all important data points before leading up the initial ServerHello response message. Yes, further fatal errors are possible on the ServerSocket after that but the exception messages are more focussed for those scenarios IMO. Yes - it won't always give the user the full picture but it certainly gives more data about the enc the server socket is being created in.

on your "preference of client or server suites" data point question... fair point - it shouldn't be a major factor in establishing a TLS connection. What about the scenario where a server has multiple key exchange algs available ? wouldn't that boolean help deduce why one key exchange was preferred over another when a hand shake might have failed ?  It's a useful data point to help diagnose why such a failure occurred IMO. No harm at least.

Reply:
> ... on your "preference of client or server suites" data point question ....
It is not expected to break the connection by changing the preference even there are multiple key exchange algs.  There may be bugs, but did you see failures caused by preference in practice?   BTW, the information is logged when ServerHello failed, and it will not be logged if the process goes to key exchange handshake message point.

> They're all important data points before leading up the initial ServerHello response message. 
I agreed that there are all important data points.  I'm just not sure if it is the right approach to dump on ServerHello failure.

Reply:
No - I haven't seen any such handshake failures as a result of the cipher suites preference. Just a theory.

OK - let's have a think about the what the best approach is for the logging frequency of such server-side data points. Perhaps another suggestion is to log such data for scenarios where the server issues a fatal alert during any part of the  handshake phase. Let me have a chat with Weibing and others and revert back.